cmd/zoekt-sourcegraph-indexserver/main.go

```diff
@@ -641,20 +641,29 @@ func sglogBranches(key string, branches []zoekt.RepositoryBranch) sglog.Field {
 }
 
 func (s *Server) indexArgs(opts IndexOptions) *indexArgs {
-	parallelism := math.Ceil(float64(s.CPUCount) / float64(s.IndexConcurrency))
+	cpuCount := s.cpuCount(opts)
+	parallelism := math.Ceil(float64(cpuCount) / float64(s.IndexConcurrency))
 	return &indexArgs{
 		IndexOptions: opts,
-
 		IndexDir:    s.IndexDir,
 		Parallelism: int(parallelism),
-
 		Incremental: true,
 
 		// 1 MB; match https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph/-/blob/cmd/symbols/internal/symbols/search.go#L22
 		FileLimit: 1 << 20,
 	}
 }
 
+// cpuCount consults both the server flags and index options to determine the number
+// of CPUs to use for indexing. If the index option is provided, it always overrides
+// the server flag.
+func (s *Server) cpuCount(opts IndexOptions) int {
+	if opts.CPUCount > 0 {
+		return int(math.Min(float64(opts.CPUCount), float64(runtime.GOMAXPROCS(0))))
```

Comment:
If a user explicitly sets a number higher than cpu count, shouldn't we allow that? EG we may find that we are not fully utilizing the CPU due to IO, and would want to "overclock" it.

any risk of using floats here misbehaving given int will round down? I'd sleep better if you just implemented min yourself.

Alternatively we upgrade to go 1.21 so we can use the new `min` builtin :) Or do something ridiculous like `slices.Min([]int{opts.CPUCount, runtime.GOMAXPROCS(0)})`

Reply:
I was following our approach for `cpu_fraction`, which is bounded between (0.0, 1.0]. But I like your idea -- for I/O intensive work, it's often helpful use more threads than CPUs and we can experiment with this. I'll rename this to `Parallelism` and let it be bigger than `runtime.GOMAXPROCS(0)`, but still upper bound it so users don't shoot themselves in the foot.

Ah the `min` stuff is a bit of a mess, will fix this :)

Reply:
Update:
* I renamed this to `ShardConcurrency` to better capture what it is -- the number of threads to use for indexing shards. Otherwise, it felt weird to say "use more CPUs than are actually available". Now we have a few different concepts (CPU fraction, shard concurrency, index concurrency), and it will be important to clean this up eventually, maybe removing CPU fraction.
* Now `ShardConcurrency` is capped at 4 * the number of available CPUs just to prevent users from shooting themselves in the foot with an accidental misconfiguration.